<!-- Google Fonts -->
<link href="https://fonts.googleapis.com/css2?family=Roboto:wght@700&display=swap" rel="stylesheet">

<div style="
    border-radius: 15px;
    border: 2px solid #2e7d32;
    padding: 20px;
    background: linear-gradient(125deg, #1b5e20 5%, #43a047 40%, #81c784 75%, #c8e6c9 100%);
    text-align: center;
    box-shadow: 0px 4px 8px rgba(0, 0, 0, 0.5);
">
    <h1 style="
        color: #fff;
        text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.7);
        font-weight: bold;
        margin-bottom: 10px;
        font-size: 36px;
        font-family: 'Roboto', sans-serif;
        letter-spacing: 1px;
    ">
        📋 Ecommerce Customer Experience 🎯
    </h1>
</div>


# 📂 Overview

* **Background** <br>
  In today’s competitive e-commerce landscape, understanding customer behavior is essential for improving **sales performance**, **customer retention**, and **service quality**. Businesses gather vast amounts of transactional and interaction data — including purchase history, payment methods, delivery times, and customer feedback. However, such data often lacks explicit labels, requiring **unsupervised learning techniques** to uncover meaningful customer patterns.
  This project focuses on analyzing multi-channel e-commerce data from a 3-month period to identify **distinct customer segments** and **key behavioral trends** that influence satisfaction and complaint likelihood.

* **Goal of the Project** <br>
  The goal is to perform **exploratory data analysis (EDA)** and apply **K-Means clustering** to group customers based on their purchasing and interaction characteristics.
  By identifying these customer segments, the business can tailor marketing strategies, optimize product offerings, and enhance service quality to better meet customer expectations.

**Key Features**

**Customer & Demographic Attributes** <br>

* `Giới tính` – Gender of the customer.
* `Độ tuổi` – Customer age.
* `Khu vực` – Regional location (North, Central, South).

**Order & Transaction Attributes** <br>

* `SL mua` – Quantity of products purchased.
* `Tổng giá trị` – Total order value.
* `PT thanh toán` – Payment method (e.g., Chuyển khoản, Tiền mặt, Ví điện tử).
* `TG giao hàng (ngày)` – Delivery time in days.
* `Kênh bán` – Sales channel (e.g., Shopee, Lazada, Website, TikTok).
* `Mã chiến dịch` – Marketing campaign identifier.

**Customer Experience Attributes** <br>

* `Đánh giá sao` – Customer satisfaction rating (1–5 stars).
* `Số lượt chat` – Number of customer service interactions.
* `Khiếu nại` – Whether the customer filed a complaint.

**Dataset Overview**

* **File Name:** `Chapter_2_Dataset_TDGroup_3 months.xlsx`
* **Time Frame:** 3 months of multi-channel transaction data.
* **Records:** Each row represents an individual order with related customer, product, and interaction details.
* **Nature of Data:** Unlabeled transactional dataset suitable for unsupervised learning and behavioral segmentation.

**Project Objective**

The primary objective is to **perform EDA and apply unsupervised clustering (K-Means)** to uncover hidden customer groups and actionable business insights.

Key components of the approach include:

**Exploratory Data Analysis (EDA):**

* Examine distributions of customer demographics, order characteristics, and satisfaction scores.
* Explore relationships between **payment type**, **channel**, and **customer rating**.
* Identify potential **outliers**, **patterns in complaints**, and **behavioral differences across regions**.
* Visualize trends in sales volume and order value across time and campaigns.

**Feature Engineering:**

* Aggregate data at **customer level** (e.g., average order value, mean rating, total orders).
* Encode categorical variables such as `Kênh bán`, `PT thanh toán`, and `Khu vực`.
* Standardize numerical features to ensure balanced clustering.
* Create derived features like:

  * Average delivery time per customer
  * Complaint ratio or satisfaction score per customer

**Clustering Modeling:**

* Apply **K-Means clustering** to identify distinct customer segments.
* Determine optimal cluster count using **Elbow Method** and **Silhouette Score**.
* Interpret each cluster’s profile (e.g., *“High-Value Loyalists”*, *“Discount Seekers”*, *“At-Risk Customers”*).
* Visualize cluster separation using **PCA** or **t-SNE**.

**Evaluation & Insights:**

* Evaluate cluster compactness and separation metrics.
* Analyze segment-specific behavior across demographics, channels, and satisfaction.
* Derive actionable insights for:

  * Marketing personalization
  * Customer service prioritization
  * Targeted retention campaigns

**Expected Outcome**

By combining EDA and clustering, the project aims to produce:

* Clear segmentation of customer groups based on transactional and behavioral patterns.
* Visual and quantitative understanding of factors driving customer satisfaction and complaints.
* Strategic insights to support **data-driven marketing** and **customer experience optimization**

<!-- Include Google Fonts for a modern font -->
<link href="https://fonts.googleapis.com/css2?family=Roboto:wght@700&display=swap" rel="stylesheet">

# <span style="color:transparent;">Import Libraries</span>

<div style="
    border-radius: 15px;
    border: 2px solid #2e7d32;
    padding: 20px;
    background: linear-gradient(125deg, #1b5e20 5%, #43a047 40%, #81c784 75%, #c8e6c9 100%);
    text-align: center;
    box-shadow: 0px 4px 8px rgba(0, 0, 0, 0.5);
">
    <h1 style="
        color: #fff;
        text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.7);
        font-weight: bold;
        margin-bottom: 10px;
        font-size: 36px;
        font-family: 'Roboto', sans-serif;
        letter-spacing: 1px;
    ">
        Import Libraries
    </h1>
</div>


In [1]:
# Core data manipulation libraries
import pandas as pd
import numpy as np

# Visualization libraries
import matplotlib.pyplot as plt
import seaborn as sns

# Statistical functions
from scipy.stats import skew

# Display utilities for Jupyter notebooks
from IPython.display import display

# Machine learning preprocessing and modeling
from sklearn.ensemble import IsolationForest
from sklearn.cluster import DBSCAN
from sklearn.decomposition import PCA
from sklearn.cluster import KMeans
from sklearn.neighbors import LocalOutlierFactor
from kneed import KneeLocator

from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer

# Suppress warnings for cleaner output
import warnings
warnings.filterwarnings("ignore")
pd.set_option("display.max_columns", 500) # To display all the columns of dataframe
pd.set_option("max_colwidth", None) # To set the width of the column to maximum

<!-- Include Google Fonts for a modern font -->
<link href="https://fonts.googleapis.com/css2?family=Roboto:wght@700&display=swap" rel="stylesheet">

# <span style="color:transparent;">Load Data</span>

<div style="
    border-radius: 15px;
    border: 2px solid #2e7d32;
    padding: 20px;
    background: linear-gradient(125deg, #1b5e20 5%, #43a047 40%, #81c784 75%, #c8e6c9 100%);
    text-align: center;
    box-shadow: 0px 4px 8px rgba(0, 0, 0, 0.5);
">
    <h1 style="
        color: #fff;
        text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.7);
        font-weight: bold;
        margin-bottom: 10px;
        font-size: 36px;
        font-family: 'Roboto', sans-serif;
        letter-spacing: 1px;
    ">
        Load Data
    </h1>
</div>

In [5]:
# Load the datasets
df_ecommerce = pd.read_excel("Chapter_2_Dataset_TDGroup_3 months.xlsx")
# Verify shapes
print("Data Shape:", df_ecommerce.shape)

Data Shape: (3357, 24)


<!-- Include Google Fonts for a modern font -->
<link href="https://fonts.googleapis.com/css2?family=Roboto:wght@700&display=swap" rel="stylesheet">

# <span style="color:transparent;">Data Preview and Info</span>

<div style="
    border-radius: 15px;
    border: 2px solid #2e7d32;
    padding: 20px;
    background: linear-gradient(125deg, #1b5e20 5%, #43a047 40%, #81c784 75%, #c8e6c9 100%);
    text-align: center;
    box-shadow: 0px 4px 8px rgba(0, 0, 0, 0.5);
">
    <h1 style="
        color: #fff;
        text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.7);
        font-weight: bold;
        margin-bottom: 10px;
        font-size: 36px;
        font-family: 'Roboto', sans-serif;
        letter-spacing: 1px;
    ">
        Data Preview and Info
    </h1>
</div>

In [6]:
# Display few rows of each dataset
df_ecommerce.head()

,Mã đơn hàng,Ngày đặt hàng,Giờ đặt hàng,Mã KH,Tên KH,Giới tính,Độ tuổi,Khu vực,Loại SP,Tên SP,Mã SP,SL mua,Giá đơn vị,Tổng giá trị,PT thanh toán,Trạng thái,TG giao hàng (ngày),Đánh giá sao,Nội dung đánh giá,Mã NV CSKH,Số lượt chat,Khiếu nại,Kênh bán,Mã chiến dịch
0,DH00001,2025-08-05,20,KH00001,Trần Anh,Nữ,27,Bắc,Phanh,Phanh 959,SP3068,2,80000,160000,Chuyển khoản,Thành công,3,4,"Sản phẩm phanh chất lượng, hài lòng mức 4 sao",NV134,1,Không,Shopee,NaN
1,DH00002,2025-08-23,22,KH00002,Nguyễn Anh,Nam,29,Bắc,Phanh,Phanh 543,SP3330,4,350000,1400000,Chuyển khoản,Thành công,3,4,"Sản phẩm phanh chất lượng, hài lòng mức 4 sao",NV124,0,Không,Website,FLASH_SALE
2,DH00003,2025-08-17,2,KH00003,Bùi Trang,Nữ,41,Bắc,Phanh,Phanh 252,SP3591,3,200000,600000,Ví điện tử,Thành công,2,4,"Sản phẩm phanh chất lượng, hài lòng mức 4 sao",NV157,2,Không,Lazada,KM_HE2025
3,DH00004,2025-08-15,15,KH00004,Trần Phương,Nam,26,Trung,Bugi,Bugi 105,SP3594,3,350000,1050000,Ví điện tử,Thành công,3,3,"Sản phẩm bugi chất lượng, hài lòng mức 3 sao",NV104,3,Không,Shopee,NaN
4,DH00005,2025-09-07,6,KH00005,Trần Thu,Nữ,39,Nam,Ắc quy,Ắc quy 556,SP2700,3,120000,360000,Chuyển khoản,Thành công,3,3,"Sản phẩm ắc quy chất lượng, hài lòng mức 3 sao",NV122,1,Có,TikTok,BACK2SCHOOL


In [7]:
# Replace space to under score.
df_ecommerce.columns = df_ecommerce.columns.str.strip().str.replace(" ", "_")

# Display information about the DataFrames
print("Data Info:")
df_ecommerce.info()

Data Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3357 entries, 0 to 3356
Data columns (total 24 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Mã_đơn_hàng          3357 non-null   object
 1   Ngày_đặt_hàng        3357 non-null   object
 2   Giờ_đặt_hàng         3357 non-null   int64 
 3   Mã_KH                3357 non-null   object
 4   Tên_KH               3357 non-null   object
 5   Giới_tính            3357 non-null   object
 6   Độ_tuổi              3357 non-null   int64 
 7   Khu_vực              3357 non-null   object
 8   Loại_SP              3357 non-null   object
 9   Tên_SP               3357 non-null   object
 10  Mã_SP                3357 non-null   object
 11  SL_mua               3357 non-null   int64 
 12  Giá_đơn_vị           3357 non-null   int64 
 13  Tổng_giá_trị         3357 non-null   int64 
 14  PT_thanh_toán        3357 non-null   object
 15  Trạng_thái           3357 non-null   object


## Insights from Initial Data Exploration

**Dataset Size and Structure**

* The dataset contains **3357 samples** and **24 features**, combining **transaction details**, **customer demographics**, **payment methods**, and **service interactions**.

**Feature Overview**

* **Numerical features:**

  * `Giờ_đặt_hàng`, `Độ_tuổi`, `SL_mua`, `Giá_đơn_vị`, `Tổng_giá_trị`, `TG_giao_hàng_(ngày)`
    → Continuous variables for scaling and clustering.

* **Categorical features:**

  * `Giới_tính`, `Khu_vực`, `Loại_SP`, `Kênh_bán`, `PT_thanh_toán`, `Trạng_thái`, `Khiếu_nại`, `Mã_chiến_dịch`, `Đánh_giá_sao`, `Số_lượt_chat`

* **Temporal feature:**

  * `Ngày_đặt_hàng` → convert to `datetime` for time-based trend analysis.

**Data Consistency and Completeness**

* Most fields are complete; only `Mã_chiến_dịch` has ~25% missing values (orders not tied to campaigns).
* Data types are appropriate: numerical as `int64`, categorical as `object`.

**Preprocessing Plan**

* Convert `Ngày_đặt_hàng` → datetime.
* Encode categorical features (`One-Hot` or `Ordinal` where applicable).
* Encode `Đánh_giá_sao` and `Số_lượt_chat` as ordinal.
* Scale numerical columns using `StandardScaler` before K-Means.
* Fill missing `Mã_chiến_dịch` with `"No Campaign"`.